storage/src/key_value/agent/block.rs

```diff
@@ -0,0 +1,364 @@
+            BlockStatus::Committed(_) | BlockStatus::Uncommitted => return Ok(()),
+        }
+        let mut transaction_serial_numbers = Vec::with_capacity(block.transactions.len());
+        let mut transaction_commitments = Vec::with_capacity(block.transactions.len());
+        let mut transaction_memos = Vec::with_capacity(block.transactions.len());
+
+        for transaction in &block.transactions {
+            transaction_serial_numbers.extend_from_slice(&transaction.old_serial_numbers[..]);
+            transaction_commitments.extend_from_slice(&transaction.new_commitments[..]);
+            transaction_memos.push(&transaction.memorandum);
+        }
+
+        if has_duplicates(transaction_serial_numbers) {
+            return Err(StorageError::DuplicateSn.into());
+        }
+        if has_duplicates(transaction_commitments) {
+            return Err(StorageError::DuplicateCm.into());
+        }
+        if has_duplicates(transaction_memos) {
+            return Err(StorageError::DuplicateMemo.into());
+        }
+
+        let mut header = vec![];
+        block.header.write(&mut header)?;
+        self.inner.store(KeyValueColumn::BlockHeader, &hash[..], &header[..])?;
+
+        let mut transactions = vec![];
+        block.write_transactions(&mut transactions)?;
+        self.inner
+            .store(KeyValueColumn::BlockTransactions, &hash[..], &transactions[..])?;
+
+        let mut child_hashes = self.get_child_block_hashes(&block.header.previous_block_hash)?;
+
+        if !child_hashes.contains(&hash) {
+            child_hashes.push(hash.clone());
+
+            self.update_child_block_hashes(&block.header.previous_block_hash, &child_hashes[..])?;
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn delete_block(&mut self, hash: &Digest) -> Result<()> {
+        match self.get_block_state(&hash)? {
+            BlockStatus::Unknown => return Ok(()),
+            BlockStatus::Uncommitted => (),
+            BlockStatus::Committed(_) => return Err(anyhow!("attempted to delete committed block")),
+        }
+
+        let header = self.get_block_header(hash)?;
+
+        self.inner.delete(KeyValueColumn::BlockHeader, &hash[..])?;
+
+        self.inner.delete(KeyValueColumn::BlockTransactions, &hash[..])?;
+
+        let mut child_hashes = self.get_child_block_hashes(&header.previous_block_hash)?;
+
+        if let Some(index) = child_hashes.iter().position(|x| x == hash) {
+            child_hashes.remove(index);
+
+            self.update_child_block_hashes(&header.previous_block_hash, &child_hashes[..])?;
+        }
+
+        Ok(())
+    }
+
+    pub(super) fn get_block_hash(&mut self, block_num: u32) -> Result<Option<Digest>> {
+        let hash = self
+            .inner
+            .get(KeyValueColumn::BlockIndex, &block_num.to_le_bytes()[..])?
+            .map(|x| x[..].into());
+        Ok(hash)
+    }
+
+    fn get_block_hash_guarded(&mut self, block_num: u32) -> Result<Digest> {
```

Comment:
if I understand correctly, the guarding aspect is related to the fact that the hash is expected to be available? don't know if the name `_guarded` doesn't suggest some locking going on, perhaps `get_block_hash_or_fail` would be a more neutral name pick?

Reply:
also, the other methods here all seem to return a `Result<T>` without an `Option`, so perhaps this method should be named `get_block_hash`, and the one called that way now should instead be `get_block_hash_relaxed` or something to that effect?